[Doc] Logging: Add Fluent Bit sidecar and Grafana Loki to Persist Ray logs (#49038)
ray-project/ray: doc/source/cluster/kubernetes/user-guides/persist-kuberay-custom-resource-logs.md

(kuberay-fluentbit)=
## Setting up logging sidecars with Fluent Bit
In this section, we give an example of how to set up log-emitting
[Fluent Bit][FluentBit] sidecars for Ray pods to send logs to [Grafana Loki][GrafanaLoki], enabling centralized log management and querying.

See the full config for a single-pod RayCluster with a logging sidecar [here][ConfigLink].
We now discuss this configuration and show how to deploy it.

### Deploy Loki monolithic mode

Follow [Deploy Loki monolithic mode](deploy-loki-monolithic-mode) to deploy Grafana Loki in monolithic mode.

### Deploy Grafana

Follow [Deploy Grafana](deploy-grafana) to set up Grafana Loki datasource and deploy Grafana.

### Configuring log processing
The first step is to create a ConfigMap with configuration
for Fluent Bit.

The following ConfigMap example configures a Fluent Bit sidecar to:
* Tail Ray logs.

> * Sends logs to a Grafana Loki endpoint.

Reviewer: ```suggestion
* Send logs to a Grafana Loki endpoint.
```

> * Add metadata to the logs for filtering by labels (e.g., `RayCluster`).

Reviewer: ```suggestion
* Add metadata to the logs for filtering by labels, for example, `RayCluster`.
```

```{literalinclude} ../configs/ray-cluster.log.yaml
:language: yaml
:start-after: Fluent Bit ConfigMap
:end-before: ---
```
A few notes on the above config:
- The `Path_Key true` line above ensures that file names are included in the log records
  emitted by Fluent Bit.
- The `Refresh_Interval 5` line asks Fluent Bit to refresh the list of files
  in the log directory once per 5 seconds, rather than the default 60.
  The reason is that the directory `/tmp/ray/session_latest/logs/` does not exist
  initially (Ray must create it first). Setting the `Refresh_Interval` low allows us to see logs
  in the Fluent Bit container's stdout sooner.

> - The `POD_LABELS` variable used in the `FILTER` section is populated using the [Kubernetes downward API][KubernetesDownwardAPI]. The label is pulled from the pod's metadata label `ray.io/cluster`, which is defined in the Fluent Bit sidecar container's environment.

Reviewer: ```suggestion
- The [Kubernetes downward API][KubernetesDownwardAPI] populates the `POD_LABELS` variable used in the `FILTER` section. It pulls the label from the pod's metadata label `ray.io/cluster`, which is defined in the Fluent Bit sidecar container's environment.
```

> - The `tenant_id` field allows you to assign logs to different tenants. In this example, the logs are sent to the `test` tenant. This configuration can be adjusted to match the tenant ID set up in your Grafana Loki instance, enabling multi-tenancy support in Grafana.

Reviewer: ```suggestion
- The `tenant_id` field allows you to assign logs to different tenants. In this example, Fluent Bit sidecar sends the logs to the `test` tenant. You can adjust this configuration to match the tenant ID set up in your Grafana Loki instance, enabling multi-tenancy support in Grafana.
```

(kuberay-fluentbit)=
## Setting up logging sidecars with Fluent Bit
In this section, we give an example of how to set up log-emitting
[Fluent Bit][FluentBit] sidecars for Ray pods to send logs to [Grafana Loki][GrafanaLoki], enabling centralized log management and querying.

See the full config for a single-pod RayCluster with a logging sidecar [here][ConfigLink].
We now discuss this configuration and show how to deploy it.

### Deploy Loki monolithic mode

Follow [Deploy Loki monolithic mode](deploy-loki-monolithic-mode) to deploy Grafana Loki in monolithic mode.

### Deploy Grafana

Follow [Deploy Grafana](deploy-grafana) to set up Grafana Loki datasource and deploy Grafana.

### Configuring log processing
The first step is to create a ConfigMap with configuration
for Fluent Bit.

The following ConfigMap example configures a Fluent Bit sidecar to:
* Tail Ray logs.
* Sends logs to a Grafana Loki endpoint.

> * Assign logs to a specific tenant using the `Tenant_ID` field, enabling multi-tenancy in Grafana Loki.(In this example, it matches the `test` tenant specified in the Grafana Loki datasource configuration.)

Reviewer 2: ```suggestion
* Assign logs to a specific tenant using the `Tenant_ID` field, enabling multi-tenancy in Grafana Loki. In this example, it matches the `test` tenant specified in the Grafana Loki datasource configuration.
```

(kuberay-logging-tldr)=
### Deploying a RayCluster with logging sidecar

To deploy the configuration described above, deploy the KubeRay Operator if you haven't yet:
Refer to the {ref}`Getting Started guide <kuberay-operator-deploy>`
for instructions on this step.

Now, run the following commands to deploy the Fluent Bit ConfigMap and a single-pod RayCluster with
a Fluent Bit sidecar.
```shell

> kubectl apply -f https://raw.githubusercontent.com/ray-project/ray/refs/heads/master/doc/source/cluster/kubernetes/configs/ray-cluster.log.yaml

Reviewer 2: delete the YAML in Ray repo and add a YAML in KubeRay repo instead

```

To access Grafana from your local machine, set up port forwarding by running:
```shell
export POD_NAME=$(kubectl get pods --namespace default -l "app.kubernetes.io/name=grafana,app.kubernetes.io/instance=grafana" -o jsonpath="{.items[0].metadata.name}")
kubectl --namespace default port-forward $POD_NAME 3000
```

This command makes Grafana available locally at `http://localhost:3000`.

* Username: "admin"
* Password: Get the password using the following command:

```shell
kubectl get secret --namespace default grafana -o jsonpath="{.data.admin-password}" | base64 --decode ; echo
```

> Finally, use a LogQL query to view logs for a specific RayCluster or RayJob, and filter by the RayCluster_name:

Reviewer 2: ```suggestion
Finally, use a LogQL query to view logs for a specific RayCluster or RayJob, and filter by the `RayCluster_name`:
```
